src/pybind/mgr/balancer/module.py

```diff
@@ -650,7 +669,10 @@ def serve(self):
                     pool_name_by_id = dict((p['pool'], p['pool_name']) for p in osdmap.dump().get('pools', []))
                     final = [int(p) for p in final]
                     final = [pool_name_by_id[p] for p in final if p in pool_name_by_id]
-                plan = self.plan_create(name, osdmap, final)
+                if self.get_module_option('mode') == 'upmap':
+                    plan = AutoUpmapPlan(name, osdmap, final)
+                else:
+                    plan = self.plan_create(name, osdmap, final)
```

Comment:
can we make this a bit cleaner but either making a corresponding CompatSetPlan, and/or making plan_create do the type branch, or possibly even a superclass that defines the common fields?

Reply:
you could also pass mode to MappingState() and initialize fields based on mode